CTF101 crypto writeups (#14)
stellarvector/blog: _writeups/2023/ctf101/encoding-challenges.md

---

> layout: writeup

Reviewer: ```suggestion
```
`mathjax` is based on `writeup` (otherwise the layouts would be broken)

parent: CTF101
grand_parent: 2023
author: r98inver
category: crypto
tags: base16 base64
title: "Encoding Challenges"
subtitle: ""
write_date: 2023-10-20
last_edit_date:
layout: mathjax # Uncomment this line to enable MathJax
---

> The first two crypto challenges of **CTF1010** where about *encoding*. 

Reviewer: ```suggestion
The first two crypto challenges of **CTF101** were about *encoding*. 
```

## sv-encoder

> We are given a python script encoding the flag: 

Reviewer: ```suggestion
We are given a Python script encoding the flag: 
```

```python
import base64 as b64

flag = b'sv{FAKE_FLAG}'
flag = b64.b16encode(flag)
for i in range(20):
	flag = b64.b64encode(flag)
for j in range(5):
	flag = b64.b16encode(flag)
flag = flag.decode()
with open('out.txt', 'w') as fh:
	fh.write(flag)
```

> As we can see, the flag is encoding with `base16`, then `base64` 20 times, then again 5 times `base16`. We are also give the output in `out.txt`. The solution consists in reversing the process, step by step:

Reviewer: ```suggestion
As we can see, the flag is encoding with `base16`, then `base64` 20 times, then again 5 times `base16`. We are also given the output in `out.txt`. The solution consists of reversing the process, step by step:
```

```python
import base64 as b64

with open('out.txt') as fh:
	flag = fh.read().strip()
flag = flag.encode()
for j in range(5):
	flag = b64.b16decode(flag)
for i in range(20):
	flag = b64.b64decode(flag)
flag = b64.b16decode(flag)
print(f'{flag = }')
# flag = b'sv{enc0d1ng_1s_n0t_3ncrypt10n}'
```

## sv-vs-encoder

The second challenge was a bit more involved. This time we have a custom encoding function, `magic_shuffle`:

```python
def magic_shuffle(poor_string):
	assert type(poor_string) == bytes, 'Magicians only use bytestrings'

	magic_string = b''

	for c in poor_string:
		
		mask = 0b11
		b0 = c & mask
		b1 = (c >> 2) & mask
		b2 = (c >> 4) & mask
		b3 = (c >> 6) & mask

		magic_char = bytes([bi + 32 for bi in [b1, b3, b0, b2] ])
		# print(f'{magic_char = }')

		magic_string += magic_char

	return magic_string
```

The flag is shuffled using this function and then `base16` encoded.

```python
flag = b'sv{FAKE_FLAG}'

flag = magic_shuffle(flag)
flag = b64.b16encode(flag)
```

> Let's look closer to `magic_shuffle`: for each character `c` in our string we do three things:

Reviewer: ```suggestion
Let's look closer at `magic_shuffle`: for each character `c` in our string we do three things:
```

> - first we derive values `b0` to to `b3` by using a `mask = 0b11` and a binary shift: this means that `b0` are the last two significant bits of `c`, `b1` the next two and so on; for instance, `a` in binary is `0b1100001`, and hence we would get `b0 = 0b01`, `b1 = 0b00`, `b2=0b10` and `b3 = 0b01`;

Reviewer: ```suggestion
- first we derive values `b0` to `b3` by using a `mask = 0b11` and a binary shift: this means that `b0` are the last two significant bits of `c`, `b1` the next two and so on; for instance, `a` in binary is `0b1100001`, and hence we would get `b0 = 0b01`, `b1 = 0b00`, `b2=0b10` and `b3 = 0b01`;
```

- then we add `32` to each one of the `bi`;
- finally, they are shuffled, and a bytestring made by `[b1, b3, b0, b2]` is returned; for instance, if we try to encode `b'a'` we get `b' !!"'`.

> The key observation here is that for each single byte we encode we get four bytes, and that these four bytes entirely define the starting byte. All we have to do is once again go backwards: for each tuple of four bytes we take them, we unshuffle them, we subtract 32 to each one of them and then use them one next to each other to rebuild the original byte.

Reviewer: ```suggestion
The key observation here is that for each single byte we encode, we get four bytes, and that these four bytes entirely define the starting byte. All we have to do is once again go backwards: for each tuple of four bytes we take them, we unshuffle them, we subtract 32 from each one of them and then put them one next to each other to rebuild the original byte.
```
